app/src/components/GameRoutes.tsx

```diff
@@ -15,23 +27,82 @@ function CurrentPlayerProvider(props: {
   joinCode: string
   children: React.ReactNode
 }) {
-  const { data, loading } = useCurrentPlayerQuery({
+  const currentAuth = React.useContext(CurrentAuthContext)
+
+  const [currentPlayer, setCurrentPlayer] = React.useState<
+    CurrentPlayerQuery["players"][0] | null
+  >(null)
+  const [redirectRoute, setRedirectRoute] = React.useState("")
+
+  // https://github.com/apollographql/react-apollo/issues/3505#issuecomment-568112061
+  const [skipCallback, setSkipCallback] = React.useState(false)
+
+  const [joinGame] = useJoinGameMutation()
+  const [signUp] = useLazyQuery(GameByJoinCodeDocument, {
+    variables: { joinCode: props.joinCode.toLocaleUpperCase() },
+    onCompleted: async data => {
+      if (skipCallback) {
+        return
+      }
+      if (data && data.games[0]) {
+        const registration = await joinGame({
+          variables: {
+            gameId: data.games[0].id,
+            clientUuid: clientUuid()
+          }
+        })
+        if (registration.data?.joinGame) {
+          await currentAuth.setJwtToken(registration.data.joinGame.jwt_token)
+        }
+        setSkipCallback(true)
+      } else {
+        setRedirectRoute(routes.game.root)
+      }
+    },
+    onError: _ => {
+      setRedirectRoute(routes.game.root)
+    }
+  })
+
+  const [
+    signIn,
+    { loading: loadingSignIn, called: calledSignIn }
+  ] = useLazyQuery(CurrentPlayerDocument, {
     variables: {
-      joinCode: props.joinCode,
+      joinCode: props.joinCode.toLocaleUpperCase(),
       clientUuid: clientUuid()
+    },
+    onCompleted: async data => {
+      if (data?.players[0]) {
+        setCurrentPlayer(data?.players[0])
+      }
     }
   })
 
-  if (!loading && !data?.players[0]) {
+  // sign in
+  React.useEffect(() => {
+    if (currentAuth.jwtToken) {
+      signIn()
+    }
+  }, [currentAuth.jwtToken])
```

Comment:
Two benefits...

1. this only tries to sign in IF there is a JWT - so we won't issue this query as "anonymous" anymore and depend on the exception handling for program control

2. no need for force update anymore. once the jwt is set, we'll trigger this side effect again